### PR TITLE
refactor(payment_links): Update API contract for dynamic transaction details and upgrade UI

### PR DESCRIPTION
**Pull request by Sakilmostak**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

the api contract is changed for payment link to specifically take position of the fields and how to render them on UI. 

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Tested through Postman: 

- Create a MCA(Bamboraapac):
```
{
    "connector_type": "payment_processor",
    "connector_name": "bamboraapac",
    "connector_account_details": {
        "auth_type": "SignatureKey",
        "api_key": "{{api_key}}",
        "key1": "{{key1}}",
        "api_secret": "{{api_secret}}"
    },
    "test_mode": false,
    "disabled": false,
    "payment_methods_enabled": [
        {
            "payment_method": "card",
            "payment_method_types": [
                {
                    "payment_method_type": "credit",
                    "card_networks": [
                        "Visa",
                        "Mastercard"
                    ],
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                },
                {
                    "payment_method_type": "debit",
                    "card_networks": [
                        "Visa",
                        "Mastercard"
                    ],
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                }
            ]
        }
    ],
    "metadata": {
        "city": "NY",
        "unit": "245",
        "acquirer_bin": "{{acquirer_bin}}",
        "acquirer_merchant_id": "{{acquirer_merchant_id}}"
    },
    "connector_webhook_details": {
        "merchant_secret": "{{merchant_secret}}"
    },
    "business_country": "US",
    "business_label": "default"
}
```

- Create a payment link request
```
{
    "amount": 1100,
    "currency": "USD",
    "confirm": false,
    "payment_link": true,
    "payment_link_config": {
        "logo": "https://hyperswitch.io/logos/hyperswitch.svg",
        "display_sdk_only": false,
        "transaction_details": [
            {
                "key": "sakil",
                "value": "mostak",
                "ui_configuration": {
                    "is_key_bold": true,
                    "is_value_bold": true,
                    "position": 3
                }
            },
            {
                "key": "Value to be filled",
                "value": "Dummy value is filled",
                "ui_configuration": {
                    "is_key_bold": false,
                    "is_value_bold": true
                }
            },
            {
                "key": "লোডটেস্ট কাজ করছে",
                "value": "বিভিন্ন ভাষায় লোডটেস্ট",
                "ui_configuration": {
                    "is_key_bold": true,
                    "is_value_bold": false,
                    "position": 4
                }
            },
            {
                "key": "Business Country",
                "value": "Germany" 
            },
            {
                "key": "Policy Number",
                "value": "938478327648",
                "ui_configuration": {
                    "is_key_bold": false,
                    "is_value_bold": false,
                    "position": 2
                }
            }
        ]
    },
    "capture_method": "manual",
    "capture_on": "2022-09-10T10:11:12Z",
    "customer_id": "StripeCustomer",
    "email": "abcdef123@gmail.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+65",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
    "setup_future_usage": "on_session",
    "browser_info": {
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "language": "nl-NL",
        "color_depth": 24,
        "screen_height": 723,
        "screen_width": 1536,
        "time_zone": 0,
        "java_enabled": true,
        "java_script_enabled": true,
        "ip_address": "127.0.0.1"
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "John",
            "last_name": "Doe"
        }
    },
    "order_details": [
        {
            "product_name": "Tea1",
            "quantity": 1,
            "amount": 110,
            "product_img_link": "https://thumbs.dreamstime.com/b/indian-tea-spices-masala-chai-33827904.jpg"
        },
        {
            "product_name": "Tea2",
            "quantity": 20,
            "amount": 110,
            "product_img_link": "https://thumbs.dreamstime.com/b/indian-tea-spices-masala-chai-33827904.jpg"
        },
        {
            "product_name": "Tea3",
            "quantity": 1,
            "amount": 110,
            "product_img_link": "https://thumbs.dreamstime.com/b/indian-tea-spices-masala-chai-33827904.jpg"
        },
         {
            "product_name": "Tea4",
            "quantity": 1,
            "amount": 110,
            "product_img_link": "https://thumbs.dreamstime.com/b/indian-tea-spices-masala-chai-33827904.jpg"
        },
        {
            "product_name": "Tea5",
            "quantity": 20,
            "amount": 110,
            "product_img_link": "https://thumbs.dreamstime.com/b/indian-tea-spices-masala-chai-33827904.jpg"
        },
        {
            "product_name": "Tea6",
            "quantity": 1,
            "amount": 110,
            "product_img_link": "https://thumbs.dreamstime.com/b/indian-tea-spices-masala-chai-33827904.jpg"
        },
         {
            "product_name": "Tea7",
            "quantity": 20,
            "amount": 110,
            "product_img_link": "https://thumbs.dreamstime.com/b/indian-tea-spices-masala-chai-33827904.jpg"
        },
        {
            "product_name": "Tea8",
            "quantity": 1,
            "amount": 110,
            "product_img_link": "https://thumbs.dreamstime.com/b/indian-tea-spices-masala-chai-33827904.jpg"
        },
         {
            "product_name": "Tea9",
            "quantity": 20,
            "amount": 110,
            "product_img_link": "https://thumbs.dreamstime.com/b/indian-tea-spices-masala-chai-33827904.jpg"
        },
        {
            "product_name": "Tea0",
            "quantity": 1,
            "amount": 110,
            "product_img_link": "https://thumbs.dreamstime.com/b/indian-tea-spices-masala-chai-33827904.jpg"
        }
    ],
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    },
    "payment_method": "card",
    "payment_method_type": "credit",
    "payment_method_data": {
        "card": {
            "card_number": "4005550000000001",
            "card_exp_month": "02",
            "card_exp_year": "26",
            "card_holder_name": "Joseph Doe",
            "card_cvc": "999"
        }
    },
    "billing": {
        "address": {
            "line1": "8th block",
            "line2": "8th block",
            "line3": "8th block",
            "city": "Bengaluru",
            "state": "Karnataka",
            "zip": "560095",
            "country": "IN",
            "first_name": "Sakil",
            "last_name": "Mostak"
        }
    }
}
```

- Redirect to the payment link, below should be the layout of the link:
image
image

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/core/payment_link/payment_link_initiate/payment_link.css [hunk 1]
```diff
@@ -96,6 +96,9 @@ body {
 
 #hyper-checkout-payment-merchant-dynamic-details {
   margin: 20px 20px 10px 20px;
+  overflow-y: scroll;
+  max-width: 35vw;
+  max-height: 10vh;
```
Comment (kashif-m) [inline crates/router/src/core/payment_link/payment_link_initiate/payment_link.css:101]: can we also add a `min-height` in pixels?

File: crates/router/src/core/payment_link/payment_link_initiate/payment_link.css [hunk 2]
```diff
@@ -714,6 +717,9 @@ body {
     flex-flow: column;
     flex-direction: column-reverse;
     margin: 0;
+    max-width: 100%;
+    overflow-y: scroll;
+    max-height: 10vh;
```
Comment (kashif-m) [inline crates/router/src/core/payment_link/payment_link_initiate/payment_link.css:722]: we can also add a `min-height` in pixels here as well

File: crates/router/src/core/payment_link/payment_link_initiate/payment_link.js [hunk 3]
```diff
@@ -617,26 +617,37 @@ function renderDynamicMerchantDetails(paymentDetails) {
 }
 
 function appendMerchantDetails(paymentDetails, merchantDynamicDetails) {
-  if (!(typeof paymentDetails.transaction_details === "string" && paymentDetails.transaction_details.length > 0)) {
+  if (!(typeof paymentDetails.transaction_details === "object" && paymentDetails.transaction_details.length > 0)) {
```
Comment (kashif-m) [inline crates/router/src/core/payment_link/payment_link_initiate/payment_link.js:620]:
```suggestion
  if (!(typeof paymentDetails.transaction_details === "object" && Object.keys(paymentDetails.transaction_details).length > 0)) {
```

File: crates/router/src/core/payment_link/payment_link_initiate/payment_link.js [hunk 4]
```diff
@@ -617,26 +617,37 @@ function renderDynamicMerchantDetails(paymentDetails) {
...
-      let maxItemsInDetails = 5;
-      for (var key in merchantDetailsObject) {
+      let maxItemsInDetails = 50;
+      for (var item of merchantDetailsObject) {
         var merchantData = document.createElement("div");
         merchantData.className = "hyper-checkout-payment-merchant-dynamic-data";
-        merchantData.innerHTML = key+": "+merchantDetailsObject[key].bold();
+        // make the key and value bold if specified in the ui_configuration
+        var key = item.ui_configuration? item.ui_configuration.is_key_bold? item.key.bold() : item.key : item.key;
+        var value = item.ui_configuration? item.ui_configuration.is_value_bold? item.key.bold() : item.value : item.value;
```
Comment (kashif-m) [inline crates/router/src/core/payment_link/payment_link_initiate/payment_link.js:649]: this should be `item.value.bold()` ?

File: crates/router/src/core/payment_link/payment_link_initiate/payment_link.js [hunk 5]
```diff
@@ -617,26 +617,37 @@ function renderDynamicMerchantDetails(paymentDetails) {
...
+        merchantData.innerHTML = key+" : "+value;
```
Comment (kashif-m) [inline crates/router/src/core/payment_link/payment_link_initiate/payment_link.js:650]: please format this file

**Review by kashif-m — CHANGES_REQUESTED** [review]

**Review by sahkal — DISMISSED** [review]
core changes LGTM!

**Comment by sahkal** [comment]
Please generate openApi specs! and commit

File: crates/router/src/core/payment_link/payment_link_initiate/payment_link.js [hunk 6]
```diff
@@ -617,26 +617,54 @@ function renderDynamicMerchantDetails(paymentDetails) {
...
+    let merchantDetailsObject = paymentDetails.transaction_details;
+    // sort the merchant details based on the position
+    // if position is null, then it will be shown at the end
+    merchantDetailsObject.sort((a, b) => {
+      if (a.ui_configuration === null || a.ui_configuration.position === null)
+        return 1;
+      if (b.ui_configuration === null || b.ui_configuration.position === null)
+        return -1;
+
+      return a.ui_configuration.position - b.ui_configuration.position;
```
Comment (kashif-m) [inline crates/router/src/core/payment_link/payment_link_initiate/payment_link.js:639]: position can be optional (defined in `TransactionDetailsUiConfiguration`), we can add type checks for the behaviour to be consistent

**Review by kashif-m — DISMISSED** [review]

**Review by kashif-m — APPROVED** [review]

**Review by sahkal — APPROVED** [review]

**Pull request merged**